File: packages/native-helpers/linux-helper-ts/src/handlers/recheck-pressed-keys.ts
import { getPressedKeys } from "../evdev/monitor.js";

export async function handleRecheckPressedKeys(
  params: Record<string, unknown>,
): Promise<{ staleKeyCodes: number[] }> {
  const pressedKeyCodes = (params.pressedKeyCodes as number[]) ?? [];
  const actuallyPressed = getPressedKeys();

  // Keys that Electron thinks are pressed but evdev says are not
  const staleKeyCodes = pressedKeyCodes.filter(
    (code) => !actuallyPressed.has(code),
  );

  return { staleKeyCodes };
}

File: packages/native-helpers/linux-helper-py/linux_helper/handlers/recheck_pressed_keys.py
"""Reconcile Electron's tracked key state with OS truth."""

from ..evdev.monitor import get_pressed_keys


def handle_recheck_pressed_keys(params: dict) -> dict:
    pressed_key_codes = params.get("pressedKeyCodes", [])
    actually_pressed = get_pressed_keys()

    stale = [code for code in pressed_key_codes if code not in actually_pressed]
    return {"staleKeyCodes": stale}
